bpo-43420: Simple optimizations for Fraction's arithmetics

skirpichev opened the pull request:
making Fraction class more usable for large arguments (>~ 10**6),
with cost 10-20% for small components case.

References:
https://www.eecis.udel.edu/~saunders/courses/822/98f/collins-notes/rnarith.ps
https://gmplib.org/, e.g. https://gmplib.org/repo/gmp/file/tip/mpq/aors.c

<!-- issue-number: [bpo-43420](https://bugs.python.org/issue43420) -->
https://bugs.python.org/issue43420
<!-- /issue-number -->

rhettinger commented on Lib/fractions.py:
```diff
@@ -382,30 +382,59 @@ def reverse(b, a):
+        x1 = math.gcd(na, db)
+        x2 = math.gcd(nb, da)
+        n, d = (na//x1)*(nb//x2), (da//x2)*(db//x1)
+        if d < 0:
```
Is there a way that *_mul()* gets called with unnormalized inputs?  If not, then this test could be skipped since the *gcd()* outputs are non-negative.

skirpichev replied:
No, it's not.  But `_mul()` and `_div()` do unnormalized outputs (in sense of denominator signs), see e.g. [this](https://github.com/python/cpython/runs/2049106689) test failure.  I suspect similar for `_add()/_sub()` (I haven't proved reversed so far), but the test suite doesn't have such cases.

skirpichev replied:
Oh, no.  For `_add()/_sub()` - both `q1` and `q2` are positive for normalized inputs and `n` is positive too (as `gcd` output).  Fixing the denominator sign is not required here.

mdickinson replied:
Aren't _all_ of `da`, `db`, `x1` and `x2` strictly positive? Can you give an example of an operation that would exercise this branch? I can see why we need to make this check for `_div` (and it looks as though the test failure you link to relates to division rather than multiplication), but I can't see how this branch would be needed.

skirpichev replied:
Indeed, only `_div()` output should be normalized (here negative sign may come from `nb`).

mdickinson reviewed (commented):
The code looks correct to me (apart from the unnecessary negativity check in `_mul` that @rhettinger pointed to). I'm less convinced that this is a desirable change, but I'll comment on the issue for that.

skirpichev committed:
Merge Fraction._add/sub() to a common helper _add_sub_()

skirpichev commented on Lib/fractions.py:
```diff
@@ -416,13 +467,9 @@ def _mul(a, b):
-        x2 = math.gcd(nb, da)
-        # Resulting fraction is normalized, because gcd(na*nb, da*db) == x1*x2.
-        # Indeed, any prime power, that divides both products on the left must
-        # either (because fractions are normalized, i.e. na and db - coprime):
```
Oops, here was a typo.  Meant da.

tim-one commented on Lib/fractions.py:
```diff
@@ -380,32 +381,109 @@ def reverse(b, a):
+    #              (da//g)*(db//g)*(g//g2)    (da//g)*(db//g2)
+    #
+    # is a normalized fraction.  This is useful because the unnormalized
+    # denominator in g2 could be much larger than g.
```
I don't understand the sentence above - like, what does "the unnormalized denominator in g2" mean? `g2` is an integer.

skirpichev replied:
It's denominator = (da//g)*(db//g)*g, i.e. an unnormalized denominator of the expression on the top.  Probably, I should reword that part.

skirpichev replied:
I hope, now is better.

skirpichev committed:
bpo-43420: Simple optimizations for Fraction's arithmetics

making Fraction class more usable for large arguments (>~ 10**6),
with cost 10-20% for small components case.

Before:
$ ./python -m timeit -s 'from fractions import Fraction as F' \
-s 'a=[F(1, _**3) for _ in range(1, 1000)]' 'sum(a)'
5 loops, best of 5: 81.2 msec per loop

After:
$ ./python -m timeit -s 'from fractions import Fraction as F' \
-s 'a=[F(1, _**3) for _ in range(1, 1000)]' 'sum(a)'
10 loops, best of 5: 23 msec per loop

References:
Knuth, TAOCP, Volume 2, 4.5.1,
https://www.eecis.udel.edu/~saunders/courses/822/98f/collins-notes/rnarith.ps,
https://gmplib.org/ (e.g. https://gmplib.org/repo/gmp/file/tip/mpq/aors.c)

skirpichev committed:
Use Fraction's private attributes in arithmetic methods

That's slightly faster for small components.

Before:
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = F(6, 5)' 'a + b'
20000 loops, best of 11: 12.1 usec per loop
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = 7' 'a + b'
20000 loops, best of 11: 11.4 usec per loop

After:
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = F(6, 5)' 'a + b'
50000 loops, best of 11: 9.74 usec per loop
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = 7' 'a + b'
20000 loops, best of 11: 16.5 usec per loop

On the master:
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = F(6, 5)' 'a + b'
50000 loops, best of 11: 9.61 usec per loop
$ ./python -m timeit -r11 -s 'from fractions import Fraction as F' \
-s 'a = F(10, 3)' -s 'b = 7' 'a + b'
50000 loops, best of 11: 9.11 usec per loop

tim-one commented on Lib/fractions.py:
```diff
@@ -380,32 +384,110 @@ def reverse(b, a):
+            s = da // g
+            t = pm(na * (db // g), nb * s)
+            g2 = math.gcd(t, g)
+            return Fraction(t // g2, s * (db // g2), _normalize=False)
```
```suggestion
            if g2 == 1:
                return Fraction(t , s * db), _normalize=False)
            else:
                return Fraction(t // g2, s * (db // g2), _normalize=False)
```
Suggested for the reason `g == 1` is a special case in the enclosing block: 1 is most likely, and some work can be skipped when the gcd is 1. I'm not aware of a proof that `g2` is most often 1, but trying some millions of "random" inputs in a test driver it was. As a sanity check, I looked at gmp's implementation, and saw that it special-cases 1 here too.

skirpichev replied:
Yes, I saw that, but I don't see any validation for this case in references.

I suspect, the reason may be same as for `g == 1` case, i.e. probability of coprimes for random inputs (`t` and `gcd` seems to be independent).  I dig into the history and that indeed was [the argument](https://gmplib.org/repo/gmp/rev/177ba6188ca8) to introduce this special case.  But that's not a real proof.  Either we should include such one or find a better reference to defend this change.  Or it's ok if we just point to [aors.c](https://gmplib.org/repo/gmp/file/tip/mpq/aors.c) (that I already did in the commit)?

skirpichev replied:
> `t` and `gcd` seems to be independent

But in fact...  We can change both parts independently (e.g. alter input numerators and freeze denominators).  Thus, it seems we can use same argument as for `g == 1`.

tim-one replied:
There is no possible question about correctness here:  if `g2 == 1`, it's utterly obvious that there's no need to explicitly divide by `g2`.

So it's a probabilistic optimization, and doesn't need "proof". It depends on the data. As I've noted before, the idea that the top-level `g == 1` test is usually true is plain false in _some_ contexts (e.g., for any addition derived from non-integer floats, the gcd is always the smaller denominator - they're all exact powers of 2).

It's plenty good enough for me that, as I said, `g2` was in fact 1 in the majority of some millions of "random" input cases I tried. That's all the justification a heuristic needs, and especially in a context where the test (`g2 == 1`) is a trivial cost compared to the potential savings. It doesn't matter if `g2` has a trillion bits - that test is cheap.

skirpichev replied:
> There is no possible question about correctness

Sure, that wasn't my point.

> So it's a probabilistic optimization, and doesn't need "proof".

I don't think so.  At least we should provide an example of some contexts, where it does make sense.  Or (better) - show, that it will payoff for random inputs.

tim-one replied:
For the third :wink: time, I already tried millions of "random" inputs  I have no interest in spending more time on it, because the eventual conclusion is already obvious.

If you need a specific context, I already gave one, but didn't spell out the consequences: additions derived from non-integer floats. There `g == 1` is never true, but `g2 == 1` is always true.

    odd1 / 2**i + odd2 / 2**j  and assume 0 < i < j
    then g = 2**i
    the numerator is  odd1 * 2**(j-i) + odd2, which is necessarily odd
    so g2 = gcd(numerator, g) = gcd(odd, power_of_2) is necessarily 1

skirpichev replied:
Ok, @tim-one, are you ok with the documentation addition, that argue this change?

skirpichev replied:
BTW, [same](https://github.com/python/cpython/pull/24779#issuecomment-792227837) random, unrelated test failure.

tim-one replied:
> are you ok with the documentation addition, that argue this change?

Oh yes. They're a real help - thanks!

skirpichev replied:
Maybe then it worth special-case `g1` and `g2` for `_mul()` as well (like [this](https://bugs.python.org/issue43420#msg388212))?

Such checks, unfortunately, are costly for very small components (that was the reason I did only `g==1` optimization in `_add_sub_()`).  Same could be seen for Raymond's benchmarks in the issue thread.

tim-one replied:
> Maybe then it worth special-case `g1` and `g2` for `_mul()` as well (like [this](https://bugs.python.org/issue43420#msg388212))?

I think so, yes. There's every reason to expect these gcds will usually be 1 (indeed, for multiplying two rationals obtained from floats, they're always 1).

The overhead of running Python-level code swamps the actual arithmetic needed for rationals with small components, but so it goes. That's not "a reason", to me, to avoid good algorithms with overheads - it's a reason not to use fractions.py at all in code for which the speed of rationals with small components is important.

skirpichev replied:
Ok, done.  Comments were slightly expanded also for the case of fractions, obtained from floats.

skirpichev committed:
Rational addition/substraction: special-case g2 == 1

Co-authored-by: Tim Peters <tim.peters@gmail.com>

skirpichev committed:
Special-case for trivial gcd in Fraction._mul()/_div() as well

skirpichev committed:
Amend _add_sub_() comment for fractions, obtained from floats

skirpichev committed:
Revert "Use Fraction's private attributes in arithmetic methods"

This reverts commit 046c84e8f970306b668d62cfacbd58f65b8c6e66.

skirpichev committed:
Revert 598b3c9630 & make _add()/_sub() less nested

tim-one reviewed (approved):
It's all fine by me now. But I'll wait a few days for Raymond and Mark to chime in.

mdickinson commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+    #
+    # Assume input fractions a and b are normalized.
+    #
+    # 1) Consider addition/substraction.
```
```suggestion
    # 1) Consider addition/subtraction.
```

mdickinson commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+    # fractions are normalized.  It's also coprime with (db//g1), because
+    # common factors are removed by g1 == gcd(na, db).
+    #
+    # As for addition/substraction, we should special-case g1 == 1
```
```suggestion
    # As for addition/subtraction, we should special-case g1 == 1
```

mdickinson commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+        nb, db = b.numerator, b.denominator
+        g1 = math.gcd(na, db)
+        if g1 > 1:
+            na //= g1
```
Please could you add a test that exercises this branch? (It doesn't seem to be covered currently.)

mdickinson commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+            db //= g2
+        n, d = na * db, nb * da
+        if nb < 0:
+            n, d = -n, -d
```
This branch also seems to be missed by the test suite; a case that exercises it would be useful.

skirpichev replied:
@mdickinson, do you mean test_fractions.py?  Because this branch triggered by test_math.py (testRemainder).

skirpichev replied:
I've added such test anyway...

tim-one replied:
> @mdickinson, do you mean test_fractions.py?

I'll speak for Mark on that :wink:: sure. An advantage of leaving modules coded in Python is that other implementations, not necessarily using CPython's C code, can use them as-is too. Even for CPython, they can often be pasted as-is into older distributions. So for modules coded in Python, we really want a self-contained test collection devoted to that module. If other tests _happen_ to trigger code paths in the module in question, that's fine (and good!), but not a substitute for the module's test collection deliberately exercising all paths in the module.

mdickinson reviewed (commented):
Looks good to me in the large; a couple of spelling nitpicks, and a couple of branches that are missed by the test suite. It would be good to fix those missed branches.

mdickinson commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+            da //= g2
+            db //= g2
+        n, d = na * db, nb * da
+        if nb < 0:
```
Nitpick: I'd probably spell this as `if d < 0`, to make it even more obviously correct. (There won't be any performance benefit to using the smaller `nb` here.)

skirpichev committed:
Apply spelling fixes

Co-authored-by: Mark Dickinson <dickinsm@gmail.com>

skirpichev committed:
Restore back initial d test in _div() per Mark suggestion

skirpichev committed:
Add coverage tests (missing in test_fractions.py)

rhettinger commented on Lib/fractions.py:
```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+        na, da = a.numerator, a.denominator
+        nb, db = b.numerator, b.denominator
+        g1 = math.gcd(na, db)
+        if g1 > 1:
```
I think the "> 1" tests should be removed.  The cost of a test is in the same ballpark as the cost of a division.  The code is clearer if you just always divide.  The performance gain mostly comes from two small gcd's being faster than a single big gcd.

skirpichev replied:
>    I think the "> 1" tests should be removed.

Maybe, I didn't benchmark this closely, but...

>    The cost of a test is in the  same ballpark as the cost of a division.

This is a micro-optimization, yes.  But ">" has not same cost as "//" even in the CPython:
```
$ ./python -m timeit -s 'a = 1' -s 'b = 1' 'a > b'
5000000 loops, best of 5: 82.3 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111111111111' -s 'b = 1' 'a > b'
5000000 loops, best of 5: 86.1 nsec per loop
$ ./python -m timeit -s 'a = 11111' -s 'b = 1' 'a // b'
2000000 loops, best of 5: 120 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111' -s 'b = 1' 'a // b'
1000000 loops, best of 5: 203 nsec per loop
$ ./python -m timeit -s 'a = 1111111111111111111111111111111111111111111111' -s 'b = 1' 'a // b'
1000000 loops, best of 5: 251 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111111111122222222222222111111111111111' \
 -s 'b = 1' 'a // b'
1000000 loops, best of 5: 294 nsec per loop
```
Maybe my measurements are primitive, but there is some pattern...

> The code is clearer if you just always divide.

I tried to explain reasons for branching in the comment above code.  Careful reader might ask: why you don't include same optimizations in the `_mul` as in `_add/sub`?

On another hand, gmplib doesn't use thise optimizations (c.f. same in mpz_add/sub).  SymPy's PythonRational class - too.

So, I did my best in defending those branches and I'm fine with removing, if @tim-one has no arguments against.

>    The performance gain mostly comes from two small gcd's being faster than a single big gcd.

This is more important, yes.

tim-one replied:
> The cost of a test is in the same ballpark as the cost of a division.

? Comparing `g` to 1 takes, at worst, small constant time, independent of how many digits `g` has. But `n // 1` takes time proportional to the number of internal digits in `n` - it's a slow, indirect way of making a physical copy of `n`. Since we expect `g` to be 1 most of the time, and these are unbounded ints, the case for checking `g > 1` seems clear to me.

rhettinger replied:
I don't mind either way.

If it were just me, I would leave out the '> 1' tests because:
1) For [small inputs](https://bugs.python.org/issue43420#msg388215), the code is about 10% faster without the guards.
2) For random 100,000 bit numerators and denominators where the gcd's were equal to 1, my timings show no discernible benefit.
3) The code and tests are simpler without the guards.

tim-one replied:
That just doesn't make much sense to me. On my box just now, with the released 3.9.1:

> C:\Windows\System32>py -m timeit -s "g = 1" "g > 1"
> 10000000 loops, best of 5: 24.2 nsec per loop
>
> C:\Windows\System32>py -m timeit -s "g = 1; t = 1 << 100000" "g > 1; t // g"
> 10000 loops, best of 5: 32.4 usec per loop

So doing the useless division not only drove the per-iteration measure from 24.2 to 32.4, the _unit_ increased by a factor of 1000(!), from nanoseconds to microseconds. That's using, by construction, an exactly 100,001 bit numerator.

And, no, removing the test barely improves that relative disaster at all:

> C:\Windows\System32>py -m timeit -s "g = 1; t = 1 << 100000" "t // g"
> 10000 loops, best of 5: 32.3 usec per loop

BTW, as expected, boost the number of numerator bits by a factor of 10 (to a million+1), and per-loop expense also becomes 10 times as much; cut it by a factor 10, and similarly the per-loop expense is also cut by a factor of 10.

With respect to the BPO message you linked to, it appeared atypical: _both_ gcds were greater than 1: `(10 / 3)  * (6 / 5)`, so there was never any possibility of testing for `> 1` paying off (while in real life a gcd of 1 is probably most common, and when multiplying Fractions obtained from floats gcd=1 is certain unless one of the inputs is an even integer (edited: used to say 0, but it's broader than just that)).

skirpichev committed:
Merge branch 'master' into opt-fractions

skirpichev committed:
Update Misc/ACKS
